ctapipe/core/telescope_component.py

```diff
@@ -0,0 +1,382 @@
+            if pattern_list is not None:
+                pattern_list.attach_subarray(subarray)
+
+    @classmethod
+    def from_name(cls, name, subarray, config=None, parent=None, **kwargs):
+        """
+        Obtain an instance of a subclass via its name
+
+        Parameters
+        ----------
+        name : str
+            Name of the subclass to obtain
+        subarray: ctapipe.instrument.SubarrayDescription
+            The current subarray for this TelescopeComponent.
+        config : traitlets.loader.Config
+            Configuration specified by config file or cmdline arguments.
+            Used to set traitlet values.
+            This argument is typically only specified when using this method
+            from within a Tool.
+        parent : ctapipe.core.Tool
+            Tool executable that is calling this component.
+            Passes the correct logger and configuration to the component.
+            This argument is typically only specified when using this method
+            from within a Tool (config need not be passed if parent is used).
+        kwargs
+
+        Returns
+        -------
+        instace
+            Instance of subclass to this class
+        """
+        requested_subclass = cls.non_abstract_subclasses()[name]
+        return requested_subclass(
+            subarray=subarray, config=config, parent=parent, **kwargs
+        )
+
+
+class TelescopePatternList(UserList):
+    """
+    Representation for a list of telescope pattern tuples. This is a helper class
+    used  by the Trait TelescopeParameter as its value type
+    """
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self._lookup = None
+        self._subarray = None
+
+        for i in range(len(self)):
+            self[i] = self.single_to_pattern(self[i])
+
+    @property
+    def tel(self):
+        """access the value per telescope_id, e.g. ``param.tel[2]``"""
+        if self._lookup:
+            return self._lookup
+        else:
+            raise RuntimeError(
+                "No TelescopeParameterLookup was registered. You must "
+                "call attach_subarray() first"
+            )
+
+    @staticmethod
+    def single_to_pattern(value):
+        # make sure we only change things that are not already a
+        # pattern tuple
+        if (
+            not isinstance(value, tuple)
+            or len(value) != 3
+            or value[0] not in {"type", "id"}
+        ):
+            return ["type", "*", value]
+
+        return value
+
+    def append(self, value):
+        """Validate and then append a new value"""
+        super().append(self.single_to_pattern(value))
+
+    def attach_subarray(self, subarray):
+        """
+        Register a SubarrayDescription so that the user-specified values can be
+        looked up by tel_id. This must be done before using the ``.tel[x]`` property
+        """
+        self._subarray = subarray
+        self._lookup.attach_subarray(subarray)
```

Comment:
Does this work when in init `self._lookup = None`?

Reply:
I assume this would not necessarily be the case, but we might type-hint in `__init__`